### ISSUE TITLE
Extending JoinClause

### ISSUE THREAD

**Issue by arthurcaputo-qs** [knex/knex#5679]
Maybe this is actually already doable and I just couldn't find out how.

On our projects we deal with an ever increasing amount of boilerplate given the way the system's tables are structured.
We already extend QueryBuilder with some extra functions to cleanup the code base somewhat.
However, JoinClause became a bit of a pain point for us as joining any of our system's tables require extra clauses that don't really have anything to do with our main logic.

So basically: would it be feasible to make JoinClause extendable?

Maybe it would be possible to just import the JoinClause module and extend it using Object.assign, but I was hoping for a more canonical way of going about things, especially since we're using TypeScript (ES Modules) in our projects.

**Comment by mercmobily** [comment]
I very much doubted you waited this long for the answer. For posterity... You want to add custom methods to reduce join boilerplate. JoinClause instances are created internally by Knex during query building (see `lib/query/joinclause.js:39`), so direct extension is less convenient than augmenting QueryBuilder.

Three paths:

1. **Use helper functions:**
```typescript
function joinSystemTable(
  builder: Knex.QueryBuilder,
  tableName: string,
  alias?: string
) {
  return builder.join(tableName, function(this: Knex.JoinClause) {
    this.on('table1.id', 'table2.id')
        .andOn('table2.deleted_at', 'IS', null)  // Your standard clauses
        .andOn('table2.tenant_id', '=', knex.raw('?', [tenantId]));
  });
}

// Usage
query.select('*').from('table1')
  .modify(joinSystemTable, 'table2', 't2');
```

2) Extend QueryBuilder with custom join methods (recommended):
```typescript
declare module 'knex' {
  interface QueryBuilder {
    joinSystemTable(table: string, alias?: string): this;
  }
}

// In your Knex initialization
QueryBuilder.prototype.joinSystemTable = function(table, alias) {
  return this.join(table, function(this: Knex.JoinClause) {
    // Your standard join logic
  });
};
```

3) Monkey‑patch JoinClause (possible, but less ideal):
```typescript
import { JoinClause } from 'knex/lib/query/joinclause';

JoinClause.prototype.systemConditions = function() {
  return this
    .andOn('deleted_at', 'IS', null)
    .andOn('tenant_id', '=', currentTenantId);
};
```
